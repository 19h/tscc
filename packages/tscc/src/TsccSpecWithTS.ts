import {TsccSpec, IInputTsccSpecJSON, ITsccSpecJSON} from '@tscc/tscc-spec'
import ITsccSpecWithTS from './ITsccSpecWithTS';
import * as ts from 'typescript';
import fs = require('fs');
import path = require('path');

export class TsError extends Error {
	constructor(
		public diagnostics:ReadonlyArray<ts.Diagnostic>
	) {
		super(ts.formatDiagnostics(diagnostics, ts.createCompilerHost({})));
	}
}

export default class TsccSpecWithTS extends TsccSpec implements ITsccSpecWithTS {
	private static loadTsConfig(
		tsConfigRoot: string = process.cwd(),
	) {
		const configFileName = ts.findConfigFile(tsConfigRoot, fs.existsSync);
		const {config: json, error} = ts.readConfigFile(configFileName, ts.sys.readFile);
		if (error) {
			throw new TsError([error])
		}
		const parsedConfig = ts.parseJsonConfigFileContent(
			json, ts.sys, path.dirname(configFileName),
			{}, configFileName
		);
		if (parsedConfig.errors.length) {
			throw new TsError(parsedConfig.errors);
		}
		return {configFileName, parsedConfig};
	}
	static loadSpecWithTS(
		tsccSpecJSONOrItsPath: string | IInputTsccSpecJSON,
		tsConfigRoot: string,
		onTsccWarning: (msg: string) => void = noop
	) {
		let {tsccSpecJSON, tsccSpecJSONPath} = TsccSpecWithTS.loadSpecRaw(tsccSpecJSONOrItsPath);
		let {configFileName, parsedConfig} = TsccSpecWithTS.loadTsConfig(tsConfigRoot);
		/**
		 * Prune compiler options
		 *  - "module" to "commonjs"
		 *  - Get values of outDir, strip it, and pass it to closure compiler
		 *  - Warn when rootDir is used - it is of no use.
		 */
		const options = parsedConfig.options;
		if (options.module !== ts.ModuleKind.CommonJS) {
			onTsccWarning(`tsickle converts TypeScript modules to Closure modules via CommonJS internally.`
				+ `"module" flag is overridden to "commonjs".`);
			options.module = ts.ModuleKind.CommonJS;
		}
		if (options.outDir) {
			onTsccWarning(`tsickle does not use --outDir, it is ignored.`);
			options.outDir = null;
		}
		if (!options.importHelpers) {
			onTsccWarning(`tsickle uses its own tslib optimized for closure compiler, importHelpers option was set forcibly.`);
			options.importHelpers = true;
		}
		return new TsccSpecWithTS(tsccSpecJSON, tsccSpecJSONPath,
			parsedConfig, configFileName);
	}
	private tsCompilerHost: ts.CompilerHost = ts.createCompilerHost(this.parsedConfig.options);
	constructor(
		tsccSpec: ITsccSpecJSON,
		basePath: string,
		private parsedConfig: ts.ParsedCommandLine,
		private tsconfigPath: string
	) {
		super(tsccSpec, basePath);
	}
	getCompilerOptions() {
		return this.parsedConfig.options;
	}
	getCompilerHost() {
		return this.tsCompilerHost;
	}
	private static readonly tsTargetToCcTarget = {
		[ts.ScriptTarget.ES3]: "ECMASCRIPT3",
		[ts.ScriptTarget.ES5]: "ECMASCRIPT5_STRICT",
		[ts.ScriptTarget.ES2015]: "ECMASCRIPT_2015",
		[ts.ScriptTarget.ES2016]: "ECMASCRIPT_2016",
		[ts.ScriptTarget.ES2017]: "ECMASCRIPT_2017",
		[ts.ScriptTarget.ES2018]: "ECMASCRIPT_2018",
		[ts.ScriptTarget.ES2019]: "ECMASCRIPT_2019",
		[ts.ScriptTarget.ESNext]: "ECMASCRIPT_NEXT"
	}
	getOutputFileNames():string[] {
		return Object.keys(this.tsccSpec.modules).map(moduleName => {
			return this.absolute(this.getOutputPrefix('cc')) + moduleName + '.js';
		});
	}
	getBaseCompilerFlags() {
		const baseFlags = this.tsccSpec.compilerFlags || {};

		const language_in: string = baseFlags["language_in"] ||
			TsccSpecWithTS.tsTargetToCcTarget[this.parsedConfig.options.target];
		const language_out = baseFlags["language_out"] || "ECMASCRIPT5";
		const compilation_level = baseFlags["compilation_level"] || "ADVANCED";
		const chunk_output_path_prefix = baseFlags["chunk_output_path_prefix"] ||
			this.absolute(this.getOutputPrefix('cc'));

		const flagsMap = Object.assign({
			language_in,
			language_out,
			compilation_level,
			chunk_output_path_prefix
		}, baseFlags);
		const outFlags: string[] = [];
		for (let [key, value] of Object.entries(flagsMap)) {
			if (Array.isArray(value)) {
				for (let val of value) {
					outFlags.push('--' + key, String(val));
				}
			} else {
				outFlags.push('--' + key, String(value));
			}
		}
		return outFlags;
	}
	getAbsoluteFileNamesSet() {
		return new Set(this.parsedConfig.fileNames.map(fileName => path.resolve(fileName)));
	}
	resolveExternalModuleTypeReference(moduleName: string) {
		const resolved = ts.resolveTypeReferenceDirective(
			moduleName,
			// Following convention of Typescript source code
			path.join(path.dirname(this.tsconfigPath), '__inferred type names__.ts'),
			this.getCompilerOptions(),
			this.getCompilerHost()
		);
		if (resolved && resolved.resolvedTypeReferenceDirective &&
			resolved.resolvedTypeReferenceDirective.isExternalLibraryImport) {
			return resolved.resolvedTypeReferenceDirective.resolvedFileName;
		}
		return null;
	}
	getExternalModuleNamesToTypeReferenceMap() {
		const out = new Map();
		this.getExternalModuleNames().forEach(moduleName => {
			let typeRef = this.resolveExternalModuleTypeReference(moduleName) || moduleName;
			out.set(moduleName, typeRef);
		});
		return out;
	}
	getExternalModulesTypeReference() {
		return this.getExternalModuleNames()
			.map(this.resolveExternalModuleTypeReference, this)
			.filter(resolvedFileName => !!resolvedFileName)
	}
	getProjectHash(): string {
		return require('crypto').createHash('sha256')
			.update(
				this.basePath + JSON.stringify(this.tsccSpec) +
				this.tsconfigPath + JSON.stringify(this.parsedConfig.options)
			)
			.digest('hex');
	}
	isDebug() {
		return this.tsccSpec.debug === true;
	}
}

function noop() {}
function exit() {return process.exit(1);}

