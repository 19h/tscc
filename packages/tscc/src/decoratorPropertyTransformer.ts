/**
 * @fileoverview Decorators in TS is not compatible with Closure Compiler, as it generates
 * code that access a class' prototype property by string literals.
 * decoratorPropertyTransformer lookup occurence of such string literal property names,
 * and replaces it with appropriate `goog.reflect.objectProperty(<prop_name>, <target>)` call.
 * 
 * Tsickle may at some point implement a similar feature. Currently it only implements some
 * other kinds of tranformation that is only made to make Angular work.
 *
 * Usage of goog.reflect.objectProperty is based on the following article:
 * {@link http://closuretools.blogspot.com/2016/06/using-polymer-with-closure-compiler-part-2.html}.
 *
 * In addition, we have to prevent this DevirtualizeMethods pass of closure compiler. However, there is
 * seem to be no stable way; see
 * nocollapse does not work - {@link https://github.com/google/closure-compiler/issues/2420}
 * sinkValue prevents inlining but does not prevent devirtualization
 * {@link https://github.com/google/closure-compiler/issues/2599}
 * 
 * The pass is prevented when the property is accessed in a global scope, so we are creating
 * aaccesses of those and remove those via regex replace after the compilation. This is hacky and
 * not guaranteed to work but was the only way to make this work. Also has to be careful about
 * accessor decorators.
 */

import * as ts from 'typescript';
import {TsickleHost} from 'tsickle';
import {isVariableRequireStatement, createGoogCall} from './transformer_utils'


function findImportedVariable(sf: ts.SourceFile, moduleName: string): ts.Identifier {
	for (let stmt of sf.statements) {
		let _ = isVariableRequireStatement(stmt);
		if (!_) continue;
		if (_.importedUrl !== moduleName) continue;
		return _.newIdent
	}
}
/**
 * The transformer needs to discern "tslib" function calls (called EmitHelpers in TS),
 * but they are simply identifiers of name `__decorate` and such, all the difference
 * lies in their `emitNode` internal property. Any functionality related to this is
 * under internal and is not available in public API.
 * This function currently access Node.emitNode.flags to achieve this
 */
function identifierIsEmitHelper(ident: ts.Identifier): boolean {
	let emitNode = ident["emitNode"];
	if (emitNode === undefined) return false;
	let flags = emitNode["flags"];
	if (typeof flags !== "number") return false;
	return (flags & ts.EmitFlags.HelperName) !== 0;
}

export default function decoratorPropertyTransformer(tsickleHost: TsickleHost):
	(context: ts.TransformationContext) => ts.Transformer<ts.SourceFile> {
	return (context: ts.TransformationContext) => {
		return (sf: ts.SourceFile) => {
			// There's nothing to do when tslib was not imported to the module.
			if (!findImportedVariable(sf, 'tslib')) return sf;
			const existingGoogReflectImport = findImportedVariable(sf, 'goog:goog.reflect');
			const googReflectImport =
				existingGoogReflectImport ||
				ts.createIdentifier(`tscc_goog_reflect_injected`);
			const annotatedPropertyAccesses: ts.Statement[] = [];
			const getId = (() => {
				let counter = 1;
				return () => `tscc_global_access_name_${counter++}`
			})();
			/**
			 * Queries whether this expression is 
			 * `tslib_1.__decorate([decorators], target, prop, descriptor)`
			 * that is generated by the TS compiler, and if so, returns a new node
			 * for __decoratr(..., goog.reflect.objectProperty(prop, target), descriptor).
			 */
			const maybeTsGeneratedDecorateCall = (node: ts.Node): ts.Node => {
				if (!ts.isCallExpression(node)) return;
				const caller = node.expression;

				if (!ts.isIdentifier(caller)) return;
				if (caller.escapedText !== ts.escapeLeadingUnderscores('__decorate')) return;
				if (!identifierIsEmitHelper(caller)) return;
				// Found a candidate. Decorator helper call signature:
				// __decorate([decoratorsArray], <target>, <propertyName>, <desc>)
				let propNameLiteral = node.arguments[2];
				if (!ts.isStringLiteral(propNameLiteral)) return;
				let propName = propNameLiteral.text;

				// Create goog.reflect.objectProperty
				const target = node.arguments[1];
				const googReflectObjectProperty = ts.setTextRange(
					ts.createCall(
						ts.createPropertyAccess(
							googReflectImport,
							ts.createIdentifier('objectProperty')
						),
						undefined,
						[
							ts.createStringLiteral(propName),
							ts.getMutableClone(target)
						]
					),
					propNameLiteral
				);
				const newCallExpression = ts.createCall(
					caller, undefined, [
						node.arguments[0],
						node.arguments[1],
						googReflectObjectProperty,
						node.arguments[3]
					]
				);
				// We will create a if (false) block at the end to annotate these properties
				// with @nocollapse.
				const propertyAccess =
					ts.createPropertyAccess(
						ts.createParen(ts.getMutableClone(target)),
						ts.createIdentifier(propName)
					)
				// Annotate it
				ts.setSyntheticLeadingComments(propertyAccess, [{
					kind: ts.SyntaxKind.MultiLineCommentTrivia,
					text: "* @nocollapase ",
					pos: -1,
					end: -1,
					hasTrailingNewLine: false
				}])
				const binaryExpression = ts.createBinary(
					ts.createElementAccess(
						ts.createIdentifier("self"),
						ts.createStringLiteral(getId())
					),
					ts.createToken(ts.SyntaxKind.FirstAssignment),
					ts.createPropertyAccess(
						ts.createParen(ts.getMutableClone(target)),
						ts.createIdentifier(propName)
					)
				);
				annotatedPropertyAccesses.push(ts.createExpressionStatement(binaryExpression));
				return newCallExpression;
			};
			let foundTransformedDecorateCall = false;
			const visitor = (node: ts.Node): ts.Node => {
				let transformed = maybeTsGeneratedDecorateCall(node);
				if (transformed) {
					foundTransformedDecorateCall = true;
					return transformed;
				}
				return ts.visitEachChild(node, visitor, context);
			}
			const newSf = visitor(sf) as ts.SourceFile;
			if (!foundTransformedDecorateCall) return newSf;
			const requireReflect = ts.createVariableStatement(
				undefined,
				ts.createVariableDeclarationList(
					[
						ts.createVariableDeclaration(
							googReflectImport,
							undefined,
							createGoogCall("require", ts.createStringLiteral('goog.reflect'))
						)
					],
					tsickleHost.es5Mode ? undefined : ts.NodeFlags.Const)
			)
			const stmts = newSf.statements.slice();
			stmts.unshift(requireReflect);
			stmts.push(
				ts.createExpressionStatement(ts.createStringLiteral("__tscc_export_start__")),
				ts.createBlock(annotatedPropertyAccesses),
				ts.createExpressionStatement(ts.createStringLiteral('__tscc_export_end__'))
			);
			return ts.updateSourceFileNode(newSf, ts.setTextRange(ts.createNodeArray(stmts), newSf.statements));
		};
	};
}

