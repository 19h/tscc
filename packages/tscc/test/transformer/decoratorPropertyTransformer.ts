///<reference types="jest" />
import * as ts from 'typescript';
import {emitWithTsickle} from 'tsickle';
import TsccSpecWithTS from '../../src/spec/TsccSpecWithTS'
import decoratorPropertyTransformer from '../../src/transformer/decoratorPropertyTransformer';
import path = require('path');

describe(`decoratorPropertyTransformer`, function () {
	test(`modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc`, () => {
		const testFiles = [
			"decorator/decorates"
		];
		const testFilePaths = testFiles.map(filename => path.resolve(__dirname, '../sample/', filename + '.ts'))
		const samplesRoot = path.resolve(__dirname, '../sample');
		const {out} = emit(samplesRoot, testFilePaths);

		[...out.values()].forEach((compiled, i) => {
			expect(compiled).toMatchSnapshot(testFiles[i]);
		})

		const {out: es5Out} = emit(samplesRoot, testFilePaths, {target: ts.ScriptTarget.ES5});
		[...es5Out.values()].forEach((compiled, i) => {
			expect(compiled).toMatchSnapshot(testFiles[i] + ' - es5');
		})

		const {out: es3Out} = emit(samplesRoot, testFilePaths, {target: ts.ScriptTarget.ES3});
		[...es3Out.values()].forEach((compiled, i) => {
			expect(compiled).toMatchSnapshot(testFiles[i] + ' - es3');
		})
	})
})

function emit(tsconfigPath: string, files: string[], override: Partial<ts.CompilerOptions> = {}) {
	const {parsedConfig} = TsccSpecWithTS.loadTsConfigFromPath(tsconfigPath);
	const {options} = parsedConfig;
	Object.assign(options, override);
	const host = ts.createCompilerHost(options);
	const program = ts.createProgram(files, options, host);
	const transformerHost = {
		shouldSkipTsickleProcessing: () => false,
		shouldIgnoreWarningsForPath: () => true,
		googmodule: false,
		pathToModuleName: x => x,
		fileNameToModuleId: x => x,
		moduleResolutionHost: ts.createCompilerHost(parsedConfig.options),
		options: parsedConfig.options,
	};

	const out = new Map<string, string>();
	const {externs} = emitWithTsickle(program, transformerHost, host, options, undefined, (fileName, data) => {
		out.set(fileName, data);
	}, undefined /* cancellationtoken */, false /* emitOnlyDtsFiles */, {
			afterTs: [decoratorPropertyTransformer(transformerHost)]
		});
	return {out, externs}
}

