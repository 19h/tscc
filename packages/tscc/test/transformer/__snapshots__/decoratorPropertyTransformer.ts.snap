// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decoratorPropertyTransformer modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc: decorator/decorates - es3 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
exports.__esModule = true;
var tslib_1 = require(\\"tslib\\");
var adornClass = function (target) { return target; };
var adornProperty = function (target, prop) { };
var adornMethod = function (target, prop, desc) { return desc; };
var adornParameter = function (target, prop, paramIndex) { };
var A = /** @class */ (function () {
    function A() {
    }
    A.prototype.aMethod = function () {
    };
    Object.defineProperty(A.prototype, \\"anAccessor\\", {
        get: function () {
            return NaN;
        },
        set: function (v) {
        },
        enumerable: true,
        configurable: true
    });
    A.prototype.aMethodWithParams = function (param) {
    };
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethod\\", A.prototype));
    tslib_1.__decorate([
        adornProperty
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aProperty\\", A.prototype));
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"anAccessor\\", A.prototype));
    tslib_1.__decorate([
        tslib_1.__param(0, adornParameter)
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethodWithParams\\", A.prototype));
    A = tslib_1.__decorate([
        adornClass
    ], A);
    return A;
}());
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).aMethod;
    self[\\"tscc_global_access_name_2\\"] = (A.prototype).aProperty;
    self[\\"tscc_global_access_name_3\\"] = (A.prototype).anAccessor;
    self[\\"tscc_global_access_name_4\\"] = (A.prototype).aMethodWithParams;
}
\\"__tscc_export_end__\\";
"
`;

exports[`decoratorPropertyTransformer modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc: decorator/decorates - es5 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var tslib_1 = require(\\"tslib\\");
var adornClass = function (target) { return target; };
var adornProperty = function (target, prop) { };
var adornMethod = function (target, prop, desc) { return desc; };
var adornParameter = function (target, prop, paramIndex) { };
var A = /** @class */ (function () {
    function A() {
    }
    A.prototype.aMethod = function () {
    };
    Object.defineProperty(A.prototype, \\"anAccessor\\", {
        get: function () {
            return NaN;
        },
        set: function (v) {
        },
        enumerable: true,
        configurable: true
    });
    A.prototype.aMethodWithParams = function (param) {
    };
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethod\\", A.prototype), null);
    tslib_1.__decorate([
        adornProperty
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aProperty\\", A.prototype), void 0);
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"anAccessor\\", A.prototype), null);
    tslib_1.__decorate([
        tslib_1.__param(0, adornParameter)
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethodWithParams\\", A.prototype), null);
    A = tslib_1.__decorate([
        adornClass
    ], A);
    return A;
}());
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).aMethod;
    self[\\"tscc_global_access_name_2\\"] = (A.prototype).aProperty;
    self[\\"tscc_global_access_name_3\\"] = (A.prototype).anAccessor;
    self[\\"tscc_global_access_name_4\\"] = (A.prototype).aMethodWithParams;
}
\\"__tscc_export_end__\\";
"
`;

exports[`decoratorPropertyTransformer modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc: decorator/decorates 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const tslib_1 = require(\\"tslib\\");
const adornClass = (target) => target;
const adornProperty = (target, prop) => { };
const adornMethod = (target, prop, desc) => desc;
const adornParameter = (target, prop, paramIndex) => { };
let A = class A {
    aMethod() {
    }
    get anAccessor() {
        return NaN;
    }
    set anAccessor(v) {
    }
    aMethodWithParams(param) {
    }
};
tslib_1.__decorate([
    adornMethod
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethod\\", A.prototype), null);
tslib_1.__decorate([
    adornProperty
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aProperty\\", A.prototype), void 0);
tslib_1.__decorate([
    adornMethod
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"anAccessor\\", A.prototype), null);
tslib_1.__decorate([
    tslib_1.__param(0, adornParameter)
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethodWithParams\\", A.prototype), null);
A = tslib_1.__decorate([
    adornClass
], A);
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).aMethod;
    self[\\"tscc_global_access_name_2\\"] = (A.prototype).aProperty;
    self[\\"tscc_global_access_name_3\\"] = (A.prototype).anAccessor;
    self[\\"tscc_global_access_name_4\\"] = (A.prototype).aMethodWithParams;
}
\\"__tscc_export_end__\\";
"
`;
